### ISSUE TITLE
Test suite not detecting now-passing expected failure tests

### ISSUE THREAD

**Issue by cflee** [pybee/voc#462]
After running git bisect based on whether marking `tests.datatypes.test_str.BinaryStrOperationTests.test_floor_divide_int` as not_implemented fails with "unexpected success" (good) or "expected failure" (bad), I found that there is a regression in the test suite in 519f61c645, as part of pep8 fixes.

When a test that is listed in not_implemented starts passing, it should fail the build. What is observed instead is that a test that is listed as not_implemented and start passing still gets marked as expected failure, until it's removed from not_implemented, then it gets marked as passing. This is incorrect behaviour.

Specifically, this works:

```python
class NotImplementedToExpectedFailure:
    def run(self, result=None):
        # Override the run method to inject the "expectingFailure" marker
        # when the test case runs.
        if self._testMethodName in getattr(self, 'not_implemented', []):
            # Mark 'expecting failure' on class. It will only be applicable
            # for this specific run.
            method = getattr(self, self._testMethodName)
            wrapper = lambda *args, **kwargs: method(*args, **kwargs)
            wrapper.__unittest_expecting_failure__ = True
            setattr(self, self._testMethodName, wrapper)
        return super().run(result=result)
```

And this fails:
```python
class NotImplementedToExpectedFailure:
    def run(self, result=None):
        # Override the run method to inject the "expectingFailure" marker
        # when the test case runs.
        if self._testMethodName in getattr(self, 'not_implemented', []):
            # Mark 'expecting failure' on class. It will only be applicable
            # for this specific run.
            def wrapper(*args, **kwargs):
                return getattr(self, self._testMethodName)(*args, **kwargs)

            wrapper.__unittest_expecting_failure__ = True
            setattr(self, self._testMethodName, wrapper)
        return super().run(result=result)
```

The specific change made:
```diff
-            method = getattr(self, self._testMethodName)
-            wrapper = lambda *args, **kwargs: method(*args, **kwargs)
+            def wrapper(*args, **kwargs):
+                return getattr(self, self._testMethodName)(*args, **kwargs)
+
```

Manually reverting this change and running pytest locally to avoid the pep8 check on Travis, pytest reports 2367 xpassed (unexpected success) and 1493 xfailed (expected failure).

I don't have time right now to dig in enough to propose a fix, so I'm filing this issue first in case it's something someone else can spot in a moment :)

**Comment by freakboy3742** [comment]
I can't obviously see why this change causes the problem we're seeing; but I can definitely see the cause - the binding of method will cause completely different behavior around exception handling. 

If a spot check confirms that the 2367 tests that are xpasses are all the same sort of problem , then we should just revert this change.

**Comment by cflee** [comment]
Hmm okay, I'll look into that!

I've just pushed a branch to my fork that reverts those two lines (with `# noqa` override for flake8) to verify that there are indeed xpasses, here is the (ongoing) [travis build for cflee/voc/fix-tests](https://travis-ci.org/cflee/voc/builds/214092275).

**Comment by cflee** [comment]
This was fixed in #541, closing.
